File: app/src/main/java/com/orgzly/android/util/AppPermissions.java
package com.orgzly.android.util;


import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.support.design.widget.Snackbar;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.view.View;

import com.orgzly.R;
import com.orgzly.android.ui.CommonActivity;
import com.orgzly.android.ui.util.ActivityUtils;

public class AppPermissions {
    public static final int FOR_LOCAL_REPO = 1;
    public static final int FOR_BOOK_EXPORT = 2;
    public static final int FOR_SYNC_START = 3;

    public static boolean isNotGranted(Context context, int requestCode) {
        String permission = permissionForRequest(requestCode);

        return ContextCompat.checkSelfPermission(context, permission) != PackageManager.PERMISSION_GRANTED;
    }

    public static boolean isGrantedOrRequest(final CommonActivity activity, int requestCode) {
        String permission = permissionForRequest(requestCode);
        int rationale = rationaleForRequest(requestCode);

        if (isNotGranted(activity, requestCode)) {
            /* Should we show an explanation? */
            if (ActivityCompat.shouldShowRequestPermissionRationale(activity, permission)) {
                View view = activity.findViewById(R.id.main_content);

                activity.showSnackbar(Snackbar.make(view, rationale, MiscUtils.SNACKBAR_WITH_ACTION_DURATION)
                        .setAction(R.string.settings, new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                ActivityUtils.openAppInfoSettings(activity);
                            }
                        }));

            } else {
                /* No explanation needed -- request the permission. */
                ActivityCompat.requestPermissions(activity, new String[]{ permission }, requestCode);
            }

            return false;

        } else {
            return true;
        }
    }


    /**
     * Mapping of request code to actual permissions.
     */
    private static String permissionForRequest(int requestCode) {
        switch (requestCode) {
            case FOR_LOCAL_REPO:
                return Manifest.permission.WRITE_EXTERNAL_STORAGE;
            case FOR_BOOK_EXPORT:
                return Manifest.permission.WRITE_EXTERNAL_STORAGE;
            case FOR_SYNC_START:
                return Manifest.permission.WRITE_EXTERNAL_STORAGE;
        }

        throw new IllegalArgumentException("Unknown permission for request code " + requestCode);
    }

    /**
     * Mapping of request code to explanation.
     */
    private static int rationaleForRequest(int requestCode) {
        switch (requestCode) {
            case FOR_LOCAL_REPO:
                return R.string.permissions_rationale_for_local_repo;
            case FOR_BOOK_EXPORT:
                return R.string.permissions_rationale_for_book_export;
            case FOR_SYNC_START:
                return R.string.permissions_rationale_for_sync_start;
        }

        throw new IllegalArgumentException("Unknown rationale for request code " + requestCode);
    }
}

File: app/src/main/java/com/orgzly/android/util/AppPermissions.kt
package com.orgzly.android.util


import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.orgzly.BuildConfig
import com.orgzly.R
import com.orgzly.android.ui.CommonActivity
import com.orgzly.android.ui.showSnackbar
import com.orgzly.android.ui.util.ActivityUtils

object AppPermissions {
    private val TAG = AppPermissions::class.java.name

    @JvmStatic
    fun isGrantedOrRequest(activity: CommonActivity, requestCode: Usage): Boolean {
        val permission = permissionForRequest(requestCode)
        val rationale = rationaleForRequest(requestCode)

        val grantedOrRequested = if (!isGranted(activity, requestCode)) {
            /* Should we show an explanation? */
            if (ActivityCompat.shouldShowRequestPermissionRationale(activity, permission)) {
                activity.showSnackbar(rationale, R.string.settings) {
                    ActivityUtils.openAppInfoSettings(activity)
                }

            } else {
                /* No explanation needed -- request the permission. */
                if (BuildConfig.LOG_DEBUG) LogUtils.d(TAG, requestCode, permission, "Requesting...")
                ActivityCompat.requestPermissions(activity, arrayOf(permission), requestCode.ordinal)
            }

            false

        } else {
            true
        }

        if (BuildConfig.LOG_DEBUG) LogUtils.d(TAG, requestCode, permission, grantedOrRequested)

        return grantedOrRequested
    }

    @JvmStatic
    fun isGranted(context: Context, requestCode: Usage): Boolean {
        val permission = permissionForRequest(requestCode)

        // WRITE_EXTERNAL_STORAGE is unused in API 30 and later
        if (permission == Manifest.permission.WRITE_EXTERNAL_STORAGE && Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            if (BuildConfig.LOG_DEBUG)
                LogUtils.d(TAG, requestCode, permission, "API " + Build.VERSION.SDK_INT + ", returning true")
            return true
        }

        val isGranted = ContextCompat.checkSelfPermission(context, permission) == PackageManager.PERMISSION_GRANTED

        if (BuildConfig.LOG_DEBUG) LogUtils.d(TAG, requestCode, permission, isGranted)

        return isGranted
    }

    /** Map request code to permission. */
    private fun permissionForRequest(requestCode: Usage): String {
        return when (requestCode) {
            Usage.LOCAL_REPO -> Manifest.permission.WRITE_EXTERNAL_STORAGE
            Usage.BOOK_EXPORT -> Manifest.permission.WRITE_EXTERNAL_STORAGE
            Usage.SYNC_START -> Manifest.permission.WRITE_EXTERNAL_STORAGE
            Usage.SAVED_SEARCHES_EXPORT_IMPORT -> Manifest.permission.WRITE_EXTERNAL_STORAGE
            Usage.EXTERNAL_FILES_ACCESS -> Manifest.permission.READ_EXTERNAL_STORAGE
        }
    }

    /** Map request code to explanation. */
    private fun rationaleForRequest(requestCode: Usage): Int {
        return when (requestCode) {
            Usage.LOCAL_REPO -> R.string.permissions_rationale_for_local_repo
            Usage.BOOK_EXPORT -> R.string.permissions_rationale_for_book_export
            Usage.SYNC_START -> R.string.permissions_rationale_for_sync_start
            Usage.SAVED_SEARCHES_EXPORT_IMPORT -> R.string.storage_permissions_missing
            Usage.EXTERNAL_FILES_ACCESS -> R.string.permissions_rationale_for_external_files_access
        }
    }

    enum class Usage {
        LOCAL_REPO,
        BOOK_EXPORT,
        SYNC_START,
        SAVED_SEARCHES_EXPORT_IMPORT,
        EXTERNAL_FILES_ACCESS
    }
}
